### PR TITLE
ESYS: Crypto backend for usage of openssl's libcrypto added.

### PR DESCRIPTION
**Pull request by JuergenReppSIT**
* The crypto backend can be selected with the option
  --with-crypto={ossl,gcrypt}.The default is gcrypt.
* Only the crypto provider independent code will remain in esys_crypto.{c,h}
  The provider dependent code can be found in:
  esys_crypto_{ossl,cgrypt}.{c,h}
* The provider dependent source files are removed from the lists
  generated in the bootstrap script and are added explicitly
  to the makefile variables in Makefile.am

Signed-off-by: Juergen Repp <Juergen.Repp@sit.fraunhofer.de>

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -63,16 +63,48 @@ AC_ARG_ENABLE([esapi],
...
+            [with_crypto=gcrypt])
+
+AM_CONDITIONAL(ESYS_OSSL, test "x$with_crypto" = "xossl")
+AM_CONDITIONAL(ESYS_GCRYPT, test "x$with_crypto" = "xgcrypt")
+
+AS_IF([test "x$with_crypto" != "xgcrypt"],
+    AS_IF([test "x$with_crypto" != "xossl"],
+          AC_MSG_ERROR([Bad value for --with-crypto $with_crypto])))
+
+AS_IF([test "x$enable_esapi" != xno -a "x$with_cryptoprovider" = "xgcrypt"],
```
Comment (tstruk) [inline configure.ac]: $with_cryptoprovider? should be $with_crypto

File: configure.ac [hunk 2]
```diff
@@ -63,16 +63,48 @@ AC_ARG_ENABLE([esapi],
...
                     [AC_MSG_ERROR([Missing required library: gcrypt.])])])
-AM_CONDITIONAL([ESAPI], [test "x$enable_esapi" != xno])
+
+AS_IF([test "x$enable_esapi" != xno -a "x$with_crypto" = "xossl"],
+      [AC_CHECK_HEADER([openssl/ssl.h],
+                       [],
+                       [AC_MSG_ERROR([Missing required header: openssl/ssl.h.])])])
+
+AS_IF([test "x$enable_esapi" != xno -a "x$with_crypto" = "xossl"],
+    AC_CHECK_LIB(ssl, OPENSSL_init_ssl, [FOUND_SSL_LIB="yes"]))
```
Comment (tstruk) [inline configure.ac]: Should we require a openssl version?
EVP_MD_CTX_create() and EVP_MD_CTX_destroy() were renamed to EVP_MD_CTX_new() and EVP_MD_CTX_free() in OpenSSL 1.1.

File: src/tss2-esys/esys_iutil.c [hunk 3]
```diff
@@ -1037,7 +1039,7 @@ iesys_check_sequence_async(ESYS_CONTEXT * esys_context)
         LOG_ERROR("Esys called in bad sequence.");
         return TSS2_ESYS_RC_BAD_SEQUENCE;
     }
-    /* TODO: Check if RESUBMISSION BELONGS HERE OR IN THE FINISH METHOD. */
+//TODO: Check if RESUBMISSION BELONGS HERE OR RATHER INTO THE FINISH METHOD.
```
Comment (tstruk) [inline src/tss2-esys/esys_iutil.c]: That was changed in 0a7027c2, why do we want to change it back?

File: src/tss2-esys/esys_crypto_ossl.h [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+#include <stddef.h>
+#include "tss2_tpm2_types.h"
+#include "tss2-sys/sysapi_util.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifdef __cplusplus
```
Comment (tstruk) [inline src/tss2-esys/esys_crypto_ossl.h]: This should go to the bottom

File: src/tss2-esys/esys_crypto_ossl.c [hunk 5]
```diff
@@ -0,0 +1,1114 @@
...
+
+    if (1 != EVP_DigestFinal_ex(mycontext->hash.ossl_context, buffer, NULL)) {
+        LOG_ERROR("Ossl error.");
+        return TSS2_ESYS_RC_GENERAL_FAILURE;
+    }
+
+    LOGBLOB_TRACE(buffer, mycontext->hash.hash_len, "read hash result");
+
+    *size = mycontext->hash.hash_len;
+    // TODO SIZE CHECK memmove(buffer, cpHash, *size);
```
Comment (tstruk) [inline src/tss2-esys/esys_crypto_ossl.c]: EVP_DigestFinal_ex() returns size in the last param if it is not null.
We should just add a local_size variable, pass it to EVP_DigestFinal_ex()
and then assert that local_size == mycontext->hash.hash_len;

File: configure.ac (hunk 2 above)
Comment (JuergenReppSIT) [reply configure.ac]: EVP_MD_CTX_create and EVP_MD_CTX_destroy are still defined in openssl/evp.h for:
openssl  1.1.0f-3 
I have tested the crypto=ossl version with OpenSSL 1.1.0 and OpenSSL 1.0.2.

File: configure.ac (hunk 1 above)
Comment (JuergenReppSIT) [reply configure.ac]: You are correct. Updated.

File: src/tss2-esys/esys_iutil.c (hunk 3 above)
Comment (JuergenReppSIT) [reply src/tss2-esys/esys_iutil.c]: Sorry for changing back by mistake. Updated

File: src/tss2-esys/esys_crypto_ossl.h (hunk 4 above)
Comment (JuergenReppSIT) [reply src/tss2-esys/esys_crypto_ossl.h]: You are correct. Updated.

File: src/tss2-esys/esys_crypto_ossl.c (hunk 5 above)
Comment (JuergenReppSIT) [reply src/tss2-esys/esys_crypto_ossl.c]: You are correct. Updated.

**Pull request merged**